numpyro/distributions/continuous.py

```diff
@@ -1188,6 +1191,99 @@ def infer_shapes(
         return batch_shape, event_shape
 
 
+class CAR(Distribution):
+    arg_constraints = {
+        "loc": constraints.real_vector,
+        "alpha": constraints.open_interval(-1, 1),
+        "tau": constraints.positive,
+        "W": constraints.positive_definite,
+    }
+    support = constraints.real_vector
+    reparametrized_params = [
+        "loc",
+        "alpha",
+        "tau",
+    ]
+
+    def __init__(self, loc, alpha, tau, W, *, is_sparse=False, validate_args=None):
+        if jnp.ndim(loc) == 0:
+            (loc,) = promote_shapes(loc, shape=(1,))
+
+        self.W = W
+
+        batch_shape = lax.broadcast_shapes(
+            jnp.shape(loc)[:-1],
+            jnp.shape(alpha),
+            jnp.shape(tau),
+            jnp.shape(self.W)[:-2]
+        )
+        event_shape = jnp.shape(self.W)[-1:]
+        self.loc = promote_shapes(loc, shape=batch_shape + event_shape)
+        self.alpha, self.tau = promote_shapes(alpha, tau, shape=batch_shape)
+        self.is_sparse = is_sparse
+
+        super(CAR, self).__init__(
+            batch_shape=batch_shape,
+            event_shape=event_shape,
+            validate_args=validate_args,
+        )
+
+    def sample(self, key, sample_shape=()):
+        if self.is_sparse:
+            if not sparse.issparse(self.W):
+                assert isinstance(self.W, np.ndarray)
+                W = self.W
+            else:
+                W = sparse.csr_matrix(self.W)
+                W = np.asarray(W.todense())
+        else:
+            assert not sparse.issparse(self.W)
+            W = self.W
+        
+        D = W.sum(axis=-1, keepdims=True) * jnp.eye(W.shape[-1])
+        tau = jnp.expand_dims(self.tau, (-2, -1))
+        alpha = jnp.expand_dims(self.alpha, (-2, -1))
+        P = tau * (D - alpha * W)
+        mvn = MultivariateNormal(self.loc, precision_matrix=P)
+        return mvn.sample(key, sample_shape=sample_shape)
+
+    @validate_sample
+    def log_prob(self, value):
+        phi = value - self.loc
+
+        if self.is_sparse:
+            if not sparse.issparse(self.W):
+                assert isinstance(self.W, np.ndarray)
+            W = sparse.csr_matrix(self.W)
+
+            D = np.asarray(W.sum(axis=-1)).squeeze()
+            D_rsqrt = np.diag(D ** (-0.5))
+
+            W_scaled = D_rsqrt.dot(W.dot(D_rsqrt))
+
+            Wphi = W.dot(phi)
```

Comment:
It seems that using the later branch is much faster
```
from scipy import sparse
import numpy as np

size = 1000
a = np.zeros((size, size))
n = size // 10
a[np.random.randint(0, size, n), np.random.randint(0, size, n)] = np.random.rand(n) + 1
print(a.shape)
W = sparse.csr_matrix(a)
D = np.asarray(W.sum(axis=-1)).squeeze(-1)

%%timeit
D_rsqrt = np.diag(D ** (-0.5))
W_scaled = D_rsqrt.dot(W.dot(D_rsqrt))

%%timeit
D_rsqrt = D ** (-0.5)
W_scaled = W * (D_rsqrt * D_rsqrt[:, np.newaxis])
```
Here `W.dot(phi)` might not work (and not propagate gradients,...) if `phi` is a jax array. I think what we want is converting W to a jax sparse array then call `W @ phi` in both branches. In addition, it is beneficial to perform `np.linalg.eigvalsh(W_scaled)` in the sparse branch (because it is an expensive operator, and we would like to precompute the constant `lam`, as explained in the Stan article).

Reply:
> D_rsqrt = np.diag(D ** (-0.5))

This has some issues in the code above in the way `a` is defined because some entries of `D` are zero. We should specify in the docs that the graph definining the adjacency matrix `W` should be a single connected component to avoid this happening. The [code to deal with a disconnected component is quite specialised](https://connordonegan.github.io/geostan/reference/stan_icar.html). If users want it, we can implement it, but in the past I have just created links between the mainland and islands to avoid this happening.

Reply:
It's nice that the use of `multiply` keeps everything sparse – `dot` made it dense.

I was playing around with some sparse solvers for the line `lam`: `scipy.sparse.linalg.eigsh(W_scaled, k=W_scaled.shape[-1]-1, return_eigenvectors=False)`. But the output and the use of k aren't ideal. I couldn't get some version using [`jax.scipy.sparse.linalg.cg`](https://jax.readthedocs.io/en/latest/_autosummary/jax.scipy.sparse.linalg.cg.html) to do anything. So we'll stick to non-sparse solvers unless you think otherwise.

Do you have a preference between `np.linalg.eigvalsh(W_scaled)` and `scipy.linalg.eigvalsh(W_scaled, np.eye(W_scaled.shape[-1])`?

Reply:
Good point about the condition of D. I think you can add some checks on the sparse branch (like positive D and lam). And remove the positive definite constraint for W.

Re generate matrix: you can set nonzero entry on the first row together with some random entries, perform a + a.T, and fill diagonal by 0 as you did.

Re sparse solver: yeah, I'm not sure if there is a sparse version of it that gives all eigvals. Let's use with the dense version for now.

Difference of lam: in sparse branch, it will be a constant in every mcmc step while in nonsparse branch, it will be recomputed by jax in every mcmc step.

Re preference: It is up to you. I tend to use np operators more often than scipy operators

Reply:
Do we want to calculate `lam` at every mcmc step? I don't think we do. `W_scaled` is a constant. i don't think we want to repeat that calculation in the dense branch

Reply:
Hmm, to avoid repeating that computation (in case you think it is expensive) in the non-sparse branch you can add a check that if W_scaled is a numpy array, then use numpy operator. You can benchmark with
```
import jax

@jax.jit
def f():
  return jax.numpy.linalg.eigvalsh(W_scaled)

@jax.jit
def g():
  return np.linalg.eigvalsh(W_scaled)
```

Reply:
To clarify: I suggested to use numpy operators on the top of numpy arrays to improve the performance. The current code repeats the computations. The benchmark code is used to confirm that it is actually repeating the computation if we use jax operators on the top of numpy array.

Reply:
If you run f, you will see that it is slow (because it is repeating the computations in later f() executions). The code is not intended to compare the speed of the jax operator and the numpy operator. In g, even if you add "very slow" numpy codes, the later execution of g is still very fast.

Reply:
For my understanding, _why_ does `jax.numpy.linalg.eigvalsh(W_scaled)` repeat the calculation at every step of the mcmc when it only needs to be calculated once? And why does `np.linalg.eigvalsh(W_scaled)` not?

Reply:
It is just how jax works. The compiled code performs all jax operators so it should perform jnp...eigvalsh again. The compiled code does not perform any non-jax operators so the value obtained from np.eigvalsh will be constant. NumPyro mcmc uses jax jit under the hood. You can try:
```
@jax.jit
def f():
    print("repeat")
    return 0.
```
to see that jax skips the print statement in the later executions. Same for numpy operators.

Reply:
Actually, I don't know how to make jax avoid repeating jnp.... operators, if that's what you are looking for. I thought it is not possible and the benchmark code seems to confirm so. If you think we can avoid repeating jnp... operators, I would love to try.

Reply:
Understood! It would be nice to treat outputs of jax ops on constant arrays as constant - it's not possible now at least for eigval but let's hope that xla (jax complier) can optimize the code better in the future.